File: DynamicPrograming/csharp/ArrayStepper.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class ArrayStepper
{
    public static bool CanReach(int[] numbers)
    {
        if (numbers.Length == 0)
        {
            return true;
        }

        var memo = new Dictionary<int, bool>();
        return Dfs(0, numbers, memo);
    }

    private static bool Dfs(int index, IReadOnlyList<int> numbers, IDictionary<int, bool> memo)
    {
        if (index >= numbers.Count - 1)
        {
            return true;
        }

        if (memo.TryGetValue(index, out var cached))
        {
            return cached;
        }

        var maxStep = numbers[index];
        for (var step = 1; step <= maxStep; step++)
        {
            if (Dfs(index + step, numbers, memo))
            {
                return memo[index] = true;
            }
        }

        return memo[index] = false;
    }
}

File: DynamicPrograming/php/arrayStepper.php
<?php

declare(strict_types=1);

function arrayStepper(array $numbers): bool
{
    $memo = [];
    $length = count($numbers);

    $dfs = function (int $index) use (&$dfs, &$memo, $numbers, $length): bool {
        if ($index >= $length - 1) {
            return true;
        }
        if (array_key_exists($index, $memo)) {
            return $memo[$index];
        }

        $maxStep = $numbers[$index];
        for ($step = 1; $step <= $maxStep; $step++) {
            if ($dfs($index + $step)) {
                return $memo[$index] = true;
            }
        }

        return $memo[$index] = false;
    };

    if ($length === 0) {
        return true;
    }

    return $dfs(0);
}
